synthesis/codon/codon.go

```diff
@@ -163,45 +155,149 @@ func Optimize(aminoAcids string, codonTable Table, randomState ...int) (string,
+
+	return table.UpdateWeights(newWeights)
+}
+
+// Translate will return an amino acid sequence which the given DNA will yield
+func (table *TranslationTable) Translate(dnaSeq string) (string, error) {
+	if dnaSeq == "" {
+		return "", errEmptySequenceString
+	}
+
+	var aminoAcids strings.Builder
+	var currentCodon strings.Builder
+	translationTable := table.TranslationMap
+	startCodonTable := table.StartCodonTable
+
+	startCodonReached := false
+	for _, letter := range dnaSeq {
+		// add current nucleotide to currentCodon
+		currentCodon.WriteRune(letter)
+
+		// if current nucleotide is the third in a codon translate to aminoAcid write to aminoAcids and reset currentCodon.
+		// use start codon table for the first codon only, erroring out if an invalid start codon is provided
+		if currentCodon.Len() == 3 {
+			if startCodonReached {
+				aminoAcids.WriteString(translationTable[strings.ToUpper(currentCodon.String())])
+			} else {
+				aminoAcid, ok := startCodonTable[strings.ToUpper(currentCodon.String())]
+				if !ok {
+					return "", fmt.Errorf("start codon %q is not in start codon table %v", currentCodon.String(), startCodonTable)
+				}
+				aminoAcids.WriteString(aminoAcid)
+				startCodonReached = true
+			}
+
+			// reset codon string builder for next codon.
+			currentCodon.Reset()
+		}
+	}
+	return aminoAcids.String(), nil
+}
+
+// weightAminoAcids weights each codon in a codon table according to input string codon frequency, adding weight to
+// the given NCBI base codon table
+func weightAminoAcids(sequence string, aminoAcids []AminoAcid) []AminoAcid {
 	sequence = strings.ToUpper(sequence)
 	codonFrequencyMap := getCodonFrequency(sequence)
 
-	for aminoAcidIndex, aminoAcid := range table.AminoAcids {
+	for aminoAcidIndex, aminoAcid := range aminoAcids {
 		// apply weights to codonTable
 		for codonIndex, codon := range aminoAcid.Codons {
-			table.AminoAcids[aminoAcidIndex].Codons[codonIndex].Weight = codonFrequencyMap[codon.Triplet]
+			aminoAcids[aminoAcidIndex].Codons[codonIndex].Weight = codonFrequencyMap[codon.Triplet]
 		}
 	}
-	return table
+
+	return aminoAcids
 }
 
-// GenerateStartCodonTable returns a mapping from the start codons of a Table to their associated amino acids.
-// For our codonTable implementation, assumes that we always map to Met.
-func (table codonTable) GenerateStartCodonTable() map[string]string {
-	result := make(map[string]string)
-	for _, codon := range table.StartCodons {
-		result[codon] = "M"
+// extractCodingRegion loops through genbank data to find all CDS (coding sequences)
+func extractCodingRegion(data genbank.Genbank) ([]string, error) {
+	codingRegions := []string{}
+
+	// iterate through the features of the genbank file and if the feature is a coding region, append the sequence to the string builder
+	for _, feature := range data.Features {
+		if feature.Type == "CDS" {
+			sequence, err := feature.GetSequence()
+			if err != nil {
+				return nil, err
```

Comment:
not covered by tests.

Reply:
It will never error ([issue](https://github.com/TimothyStiles/poly/issues/352)),should I ignore the error here or make mocks?